v2.0.0

GeorgeKaraszi opened the pull request:
This is a complete refactor and restructuring of the Gem.

Deprecating Dependencies
- Hashie (Response Object Container) (Replaced with a Generic `CaptainObject` (OpenStruct))
- Faraday (Old, heavy HTTP client)
- ActiveSupport (usage of `#blank?` is useful, but shouldn't be a requirement)

In Favor of:
- Http-rb: Light weight, and really fast HTTP client written in mostly ruby with better exception handling.
- Oj - low-level JSON parser written in C; lighting fast.

## Dropping Rails loader
No longer do we auto-initialize / require the captain gem when a Rails app is present.

This is due to favoring the [the-captain-rails](https://github.com/VianetManagement/the-captain-rails) gem that will load this Gem as well as integrate other Rails specific logic.

This also keeps this gem **very** Ruby specific. Without the overhead of other outside dependencies.


---

## Requests classes (ex: `TheCaptain::User`, `TheCaptain::IPAddress`, ... etc)
 - Will now return a `TheCaptain::Response::CaptainVessel` class instance. 
 - Redesigned to allow for multiple API url paths for the same context

## `TheCaptain::Response::CaptainVessel`
- Used to parse response's from The Captain into `CaptainObject`'s
- Maintains accessors to view the raw request information
- Can verify the response status to ensure its valid or invalid.

## `TheCaptain::Response::CaptainObject`
 - Is the decoded body object that was decoded via `TheCaptain::Response::CaptainVessel`
 - Is a inherited `OpenStruct` class thats dynamically parsed and instantiated  by the `Oj` Gem.

## `TheCaptain::CaptainClient`
- Designed to make and process requests as well as standard error handling for requests.
- `.active_client` or `.default_client` Class Methods - Designed to be thread-safe. Allowing multiple threads to share a common HTTP object without the need to be re-instantiated.
- `#decode_response` - Can decode responses into the `TheCaptain::Response::CaptainVessel` klass

## Validations
- Striped from `TheCaptain::APIResource` in favor of collectively housing validation methods into the `TheCaptain::Utility::Validation` module.

## Param normalization / conversations
- Striped away from `TheCaptain::APIResource` in favor of collectively housing request param handling into the `TheCaptain::Utility::Helper` module.

## Configurations
- `raise_http_errors` (default: false)- Setting this true, will raise an exception any time the response is not within the 2XX or 3XX range. 
- Configuring via Environmental Variables
  - `CAPTAIN_API_URL`(default: https://thecaptain.elevatorup.com/v2) - Sets the base url of the API
  - `CAPTAIN_API_KEY`

---

## Exceptions
- Local/Client Issue's will always be Prefixed with `Client`
  - `TheCaptain:Error::ClientInvalidRequestError` - User tried to submit a request that wasn't: `get`, `post`, or `patch`
  - `TheCaptain::Error::ClientInvalidResourceError` - A resource was attempted to be parsed but wasn't the correct object type expected.
  - `TheCaptain::Error::ClientError` - A pre-mature action was taken [ex: decoding a response without making a request first]
  - `TheCaptain::Error::ClientAuthenticationError` - Missing API Key
- API request errors will always be Prefixed with `API`
  - `TheCaptain::Error::APIAuthorizationError` - The captain responds back stating the API key is invalid
  - `TheCaptain::Error::APIConnectionError` - Failed to establish a good connection to The Captains API.
  - `TheCaptain::Error::APIInvalidRequestError` - A request to the captain was malformed and the param's were sent or requested were incorrect.

Each exception will provide a message detailing what happened. In the case of a API exception, connection information may also be provided. This will help aid and inform the developer what happened between his client and the target server.

GeorgeKaraszi committed:
Inital commit to version 2

GeorgeKaraszi committed:
Rubocop update

GeorgeKaraszi committed:
Migrate all old code over to a legacy directory

GeorgeKaraszi committed:
Depercate old code

GeorgeKaraszi committed:
Setup new captain connection client using HTTP-RB and collectively put expections in a single area

GeorgeKaraszi committed:
ignore legacy directory

GeorgeKaraszi committed:
Remove legacy code

GeorgeKaraszi committed:
Remove old examples

GeorgeKaraszi committed:
Drop Hashie and use open struct to decode responses

GeorgeKaraszi committed:
format and drop gemspec executables

GeorgeKaraszi committed:
Enable environmental variables to be used for configuration setup

GeorgeKaraszi committed:
Allow for multiple api destinations for the same class

GeorgeKaraszi committed:
Add IP Address and Message classes

GeorgeKaraszi committed:
Set default base for missing api path

GeorgeKaraszi committed:
Introduce initial stats class with validation

GeorgeKaraszi committed:
Migrate the inital normaizing of params to a module class

GeorgeKaraszi committed:
Require needed classes

GeorgeKaraszi committed:
Refactor and reduce Gem error states

GeorgeKaraszi committed:
Update Code Climents rubocop tool

GeorgeKaraszi committed:
Fix namespace issue and client error checking issue

GeorgeKaraszi committed:
Begin testing configuation settings.. And drop looking for API version

GeorgeKaraszi committed:
Drop legacy tests

GeorgeKaraszi committed:
Rename CaptainContainer to CaptainVessel... And parse JSON objects via OJ callbacks

GeorgeKaraszi committed:
Bump dev ruby version to 2.5.1

GeorgeKaraszi committed:
Fix exception initializion issues and refactor CaptainObject down

GeorgeKaraszi committed:
Spec test response object classes

GeorgeKaraszi committed:
Bump travis ruby versions

GeorgeKaraszi committed:
Fix Utility Validation Namespace issue

GeorgeKaraszi committed:
Begin testing user class

GeorgeKaraszi committed:
Always wipe connection thread stoarage for each test

GeorgeKaraszi committed:
Begin testing validation methods and stats routes

GeorgeKaraszi committed:
Add initial content API operations

GeorgeKaraszi committed:
Require content API

GeorgeKaraszi committed:
Drop Rails initializer to favor the captains rails gem

GeorgeKaraszi commented:
Hey @altereagle  & @joshua-honig .. This isn't completed; in-fact URL's are of corse subject to change. But I'd like some review as it stands ATM.

altereagle commented:
Yep, I'll dig into this!

joshua-honig commented:
Sounds good. First impressions: I do like what I'm seeing here with the general trimming of dependencies.

GeorgeKaraszi commented:
Just wrapped up outlining exception information. More details can be provided later.

GeorgeKaraszi committed:
Add debugging inspecting method

altereagle commented on lib/the_captain/captain_client.rb:
```diff
@@ -0,0 +1,99 @@
+      @conn.post(url, form: params)
+    end
+
+    def patch(url, params = {})
```
We're probably not going to have a PATCH options anymore. We'll be talking about that today.

GeorgeKaraszi commented on lib/the_captain/ip_address.rb:
```diff
@@ -0,0 +1,30 @@
+              stats: "/stats",
+              usage: "/usage"
+
+    def retrieve(**params)
```
Semantically, its a feeling for how the code presents itself.

`retrieve_X` to me sounds like a term that houses both a singular return as well as a multi-array return. To where as `get_X`,  (again for me) sounds singular exclusively.

altereagle replied:
Semantically that makes sense to me.  It keeps you away from needing a `get_all_x` and `get_x`, if you are presuming that getting something means get *one* thing.

This doesn't work with this convention
```ruby
def get(**params)
```

This works, but its super gross
```ruby
def get(**ip_address_params)
```

I see where you are coming from.

GeorgeKaraszi committed:
Remove Patch HTTP method references

GeorgeKaraszi committed:
Drop client-side validation

To many varaibles at play to confidentiality build a client-side validation set of methods.

GeorgeKaraszi committed:
Introduce new user paths

GeorgeKaraszi committed:
Drop stats and messages

GeorgeKaraszi committed:
Refactor User and Ip Address methods

GeorgeKaraszi committed:
fix url helper bug

GeorgeKaraszi committed:
introduce content and credit cards

GeorgeKaraszi committed:
Refactor to define repeatable path methods using Meta-programming

GeorgeKaraszi committed:
fix user's related content path

GeorgeKaraszi committed:
spec test user requests

GeorgeKaraszi committed:
Drop dynamic resource_id reference key

GeorgeKaraszi committed:
spec test resource helper

GeorgeKaraszi committed:
Spec test and refactor request method testing

GeorgeKaraszi committed:
explicitly state we are defining get path methods

GeorgeKaraszi committed:
pretty'fy method check

GeorgeKaraszi committed:
Add Email Address Get roues.

GeorgeKaraszi committed:
prevent queries when disabled

GeorgeKaraszi committed:
URI escape special symbols for valid URL pattern

GeorgeKaraszi committed:
Target V3 subpath

GeorgeKaraszi committed:
enforce sending post data as json format

GeorgeKaraszi committed:
Add rubocop-rspec and fix rubocop errors

GeorgeKaraszi committed:
Add example of submition usage

GeorgeKaraszi committed:
version bump

GeorgeKaraszi committed:
delegate dig to CaptainObject

GeorgeKaraszi committed:
implement offical URL

GeorgeKaraszi committed:
ensure api info are strings

GeorgeKaraszi committed:
Merge pull request #18 from VianetManagement/3.0.0

3.0.0
